pkg/distributor/distributor.go

```diff
@@ -805,6 +806,9 @@ func (d *Distributor) prePushRelabelMiddleware(next push.Func) push.Func {
 		}
 
 		if len(removeTsIndexes) > 0 {
+			for _, removeTsIndex := range removeTsIndexes {
+				mimirpb.ReuseTimeseries(req.Timeseries[removeTsIndex].TimeSeries)
```

Comment:
I'm not sure why this wouldn't return the `yoloSlice` to the pool, `util.RemoveSliceIndexes` shouldn't replace the slice although it might shorten it by updating the length property.

Now that I think about it... if `util.RemoveSliceIndexes` removes index `0` of the slice then even though the underlying data array of the slice doesn't get replaced, the offset `0` gets shifted forward and I suspect that the positions on the underlying array which are before the new offset `0` will never be accessible again. Is this what you were referring to?

Reply:
> I'm not sure why this wouldn't return the yoloSlice to the pool, `util.RemoveSliceIndexes` shouldn't replace the slice although it might shorten it by updating the length property.

1. for elements that we're going to remove, we call `mimirpb.ReuseTimeseries(req.Timeseries[removeTsIndex].TimeSeries)` here.
2. for entire slice, we will eventually call `mimirpb.ReuseSlice(req.Timeseries)`, added as "cleanup" function in [push.go](https://github.com/grafana/mimir/blob/d3b8abee9cee98f9c723a26f5db4f0d95a6f6417/pkg/util/push/push.go#L112-L111). We perform this call on updated `req.Timeseries`, which at this point when cleanup runs only has timeseries that were actually pushed to the ingester, but not timeseries that were removed due to validation or relabeling.

Call to `mimirpb.ReuseSlice(req.Timeseries)` in step 2 will call `ReuseTimeseries(ts[i].TimeSeries)` on individual elements, and then also return their `yoloSlice` to the pool:

https://github.com/grafana/mimir/blob/70f91b8e1fc8e0442371bde89b708e1c2e2d2c90/pkg/mimirpb/timeseries.go#L290-L301

But in step 1, we only call `mimirpb.ReuseTimeseries`, but don't return `yoloSlice` to the pool. And since these elements are no longer part of `req.Timeseries` when cleanup (step 2) is done, their `yoloSlice` will not be returned back.

Reply:
You're right, I missed that. 

Fixed here: ead8758e9